### DISCUSSION TITLE
How to forward ANSI escape codes thru ripgrep?

### DISCUSSION THREAD

**Discussion by hrls** [BurntSushi/ripgrep#1920] [General] [answered]
Is there any way to forward colors through ripgrep? At least ANSI escape codes.  

As for example: I want print content of file with syntax highlighting via [bat](https://github.com/sharkdp/bat), then filter out some lines by ripgrep and print them with colors of bat (and maybe ripgrep as overlay too).

<img width="753" alt="image" src="https://user-images.githubusercontent.com/3259009/124125519-5ca73e00-da82-11eb-9589-d42150671503.png">

**Comment by BurntSushi** [comment, marked as answer]
The question here is not quite framed right. The issue here isn't whether ripgrep is "forwarding" colors or not. ripgrep prints the bytes it gets exactly as-is. If colors are emitted via ANSI escape sequences, then of course those are sent through as well. Indeed, your example isn't even showing a failure of forwarding colors. What it's showing is that your query for `main(` doesn't produce any results. Why? Well, I'm guessing it's because `bat` is emitting ANSI escape codes between `main` and `(`, and thus, `main(` does not match anything.

There's really not much ripgrep can do about this. You either have to ensure that your queries don't overlap with syntax highlighting boundaries, or otherwise ensure that your regex is robust with respect to ANSI escape codes. e.g., Maybe something like `rg 'main(?-u:.*)\('` would work here.

I think what you actually want is a way to search data that contains ANSI escape codes _as if_ those ANSI escape codes weren't there, but that when it comes time to print the results, those ANSI escape codes should be re-inserted in their original location. This doesn't really fit within ripgrep's project goals, and moreover, implementing such a thing would be quite tricky in practice.

**Reply by hrls** [reply]
> way to search data that contains ANSI escape codes as if those ANSI escape codes weren't there

Yes, this is exactly what I mean.

> This doesn't really fit within ripgrep's project goals, and moreover, implementing such a thing would be quite tricky in practice.

Agree.

`rg 'main(?-u:.*)\('` works as expected. Thank you for answer.
